File: frontend/src/api/meshProcessing.ts
import { apiClient } from "./client.js";

export interface MeshAnalysis {
  vertex_count: number;
  face_count: number;
  is_watertight: boolean;
  is_manifold: boolean;
  has_duplicate_vertices: boolean;
  bounding_box: {
    min_x: number;
    max_x: number;
    min_y: number;
    max_y: number;
    min_z: number;
    max_z: number;
  };
  file_size_bytes: number;
}

export type RepairOperation =
  | "remove_duplicates"
  | "fix_normals"
  | "fill_holes"
  | "remove_degenerate";

export interface SimplifyRequest {
  source_file: string;
  target_faces: number;
}

export interface RepairRequest {
  source_file: string;
  operations: RepairOperation[];
}

export interface ProcessingResult {
  output_file: string;
  processing: {
    operation: string;
    params: Record<string, unknown>;
    source_file: string;
    output_file: string;
    processed_at: string;
  };
}

export interface ClipFloorRequest {
  source_file: string;
  y_threshold?: number | null;
}

export interface ClipFloorResult {
  output_file: string;
  y_threshold_used: number;
  vertices_removed: number;
  faces_removed: number;
}

export interface RemoveComponentsRequest {
  source_file: string;
  min_component_ratio?: number;
}

export interface RemoveComponentsResult {
  output_file: string;
  components_found: number;
  components_removed: number;
  triangles_removed: number;
}

export async function analyzeMesh(
  assetId: string,
  sourceFile: string = "mesh.glb"
): Promise<MeshAnalysis> {
  const { data } = await apiClient.get<MeshAnalysis>(
    `/assets/${assetId}/process/analyze`,
    { params: { source_file: sourceFile } }
  );
  return data;
}

export async function simplifyMesh(
  assetId: string,
  req: SimplifyRequest
): Promise<ProcessingResult> {
  const { data } = await apiClient.post<ProcessingResult>(
    `/assets/${assetId}/process/simplify`,
    req
  );
  return data;
}

export async function repairMesh(
  assetId: string,
  req: RepairRequest
): Promise<ProcessingResult> {
  const { data } = await apiClient.post<ProcessingResult>(
    `/assets/${assetId}/process/repair`,
    req
  );
  return data;
}

export async function getMeshSources(
  assetId: string
): Promise<{ sources: string[] }> {
  const { data } = await apiClient.get<{ sources: string[] }>(
    `/assets/${assetId}/process/sources`
  );
  return data;
}

export async function clipFloor(
  assetId: string,
  req: ClipFloorRequest
): Promise<ClipFloorResult> {
  const { data } = await apiClient.post<ClipFloorResult>(
    `/assets/${assetId}/process/clip-floor`,
    req
  );
  return data;
}

export async function removeComponents(
  assetId: string,
  req: RemoveComponentsRequest
): Promise<RemoveComponentsResult> {
  const { data } = await apiClient.post<RemoveComponentsResult>(
    `/assets/${assetId}/process/remove-components`,
    req
  );
  return data;
}

File: api/app/schemas/mesh_processing.py
"""Pydantic-Schemas für Mesh-Processing-API."""

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field, field_validator


class RepairOperation(str, Enum):
    """Repair-Operationen für Mesh-Bearbeitung."""

    REMOVE_DUPLICATES = "remove_duplicates"
    FIX_NORMALS = "fix_normals"
    FILL_HOLES = "fill_holes"
    REMOVE_DEGENERATE = "remove_degenerate"


class MeshAnalysis(BaseModel):
    """Analyse-Ergebnis eines Meshes."""

    vertex_count: int
    face_count: int
    is_watertight: bool
    is_manifold: bool
    has_duplicate_vertices: bool
    bounding_box: dict[str, float] = Field(
        description="min_x, max_x, min_y, max_y, min_z, max_z"
    )
    file_size_bytes: int


class SimplifyRequest(BaseModel):
    """Request für POST /assets/{asset_id}/process/simplify."""

    source_file: str = Field(..., description="z.B. mesh.glb")
    target_faces: int = Field(..., ge=1, description="Ziel-Anzahl Faces")


class RepairRequest(BaseModel):
    """Request für POST /assets/{asset_id}/process/repair."""

    source_file: str = Field(..., description="z.B. mesh.glb")
    operations: list[RepairOperation] = Field(
        ..., min_length=1, description="Liste der Repair-Operationen"
    )


class ClipFloorRequest(BaseModel):
    """Request für POST /assets/{asset_id}/process/clip-floor."""

    source_file: str = Field(..., description="z.B. mesh.glb")
    y_threshold: float | None = Field(
        default=None,
        description="Y-Schwellwert; null = Auto-Detect (untere 5%)",
    )


class RemoveComponentsRequest(BaseModel):
    """Request für POST /assets/{asset_id}/process/remove-components."""

    source_file: str = Field(..., description="z.B. mesh.glb")
    min_component_ratio: float = Field(
        default=0.05,
        ge=0.0,
        le=1.0,
        description="Komponenten unter X% der Hauptmesh-Größe werden entfernt",
    )


class LodGenerateRequest(BaseModel):
    """Request für POST /assets/{asset_id}/lods."""

    source_file: str = Field(..., description="Quell-Mesh, z.B. mesh.glb")
    ratios: list[float] = Field(
        default=[1.0, 0.5, 0.25],
        min_length=1,
        max_length=8,
        description="Reduktionsfaktoren je LOD-Stufe (0.0–1.0). "
                    "LOD0=1.0 (Original), LOD1=0.5 (50%), LOD2=0.25 (25%)",
    )

    @field_validator("ratios")
    @classmethod
    def validate_ratios(cls, v: list[float]) -> list[float]:
        for r in v:
            if not (0.0 < r <= 1.0):
                raise ValueError(f"Alle Ratios müssen in (0, 1] liegen, erhalten: {r}")
        return v


class LodResult(BaseModel):
    """Einzelnes LOD-Ergebnis."""

    level: int = Field(description="LOD-Stufe (0 = höchste Qualität)")
    output_file: str = Field(description="Dateiname, z.B. mesh_lod0.glb")
    ratio: float = Field(description="Angewendeter Reduktionsfaktor")
    actual_faces: int = Field(description="Tatsächliche Face-Anzahl nach Decimation")


class LodGenerateResponse(BaseModel):
    """Response für POST /assets/{asset_id}/lods."""

    lods: list[LodResult]
    source_file: str


class ProcessingResult(BaseModel):
    """Response für simplify/repair."""

    output_file: str
    processing: dict[str, Any] = Field(
        description="Eintrag für metadata.json processing-Array"
    )


class TextureBakeRequest(BaseModel):
    """Request für POST /assets/{asset_id}/texture/bake."""

    source_mesh: str = Field(..., description="High-Poly mit Texturen, z.B. mesh.glb")
    target_mesh: str = Field(
        ...,
        description="Low-Poly ohne Texturen, z.B. mesh_simplified_10000.glb",
    )
    resolution: int = Field(
        default=1024,
        ge=512,
        le=2048,
        description="Textur-Auflösung: 512, 1024, 2048",
    )
    bake_types: list[str] = Field(
        default=["diffuse", "roughness", "metallic"],
        description="Bake-Typen: diffuse, roughness, metallic",
    )

    @field_validator("bake_types")
    @classmethod
    def validate_bake_types(cls, v: list[str]) -> list[str]:
        valid = {"diffuse", "roughness", "metallic"}
        filtered = [t for t in v if t in valid]
        return filtered if filtered else ["diffuse", "roughness", "metallic"]


class TextureBakeStartResponse(BaseModel):
    """Response für POST /assets/{asset_id}/texture/bake (Job gestartet)."""

    job_id: str
    status: str = "pending"


class TextureBakeStatusResponse(BaseModel):
    """Response für GET /assets/{asset_id}/texture/bake/status/{job_id}."""

    job_id: str
    status: str  # pending | processing | done | failed
    output_file: str | None = None
    duration_seconds: float | None = None
    error_msg: str | None = None
